src/common-util/src/main/java/ee/ria/xroad/common/certificateprofile/impl/GenericCertificateProfileInfoProvider.java

```diff
@@ -0,0 +1,144 @@
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package ee.ria.xroad.common.certificateprofile.impl;
+
+import ee.ria.xroad.common.CodedException;
+import ee.ria.xroad.common.ErrorCodes;
+import ee.ria.xroad.common.certificateprofile.AuthCertificateProfileInfo;
+import ee.ria.xroad.common.certificateprofile.CertificateProfileInfoProvider;
+import ee.ria.xroad.common.certificateprofile.DnFieldDescription;
+import ee.ria.xroad.common.certificateprofile.SignCertificateProfileInfo;
+import ee.ria.xroad.common.identifier.ClientId;
+import ee.ria.xroad.common.util.CertUtils;
+
+import org.bouncycastle.asn1.x500.X500Name;
+import org.bouncycastle.asn1.x500.style.BCStyle;
+
+import javax.security.auth.x500.X500Principal;
+
+import java.security.cert.X509Certificate;
+
+/**
+ * Generic certificate profile
+ */
+public class GenericCertificateProfileInfoProvider
+        implements CertificateProfileInfoProvider {
+
+    @Override
+    public AuthCertificateProfileInfo getAuthCertProfile(AuthCertificateProfileInfo.Parameters params) {
+        return new GenericAuthCertificateProfileInfo(params);
+    }
+
+    @Override
+    public SignCertificateProfileInfo getSignCertProfile(SignCertificateProfileInfo.Parameters params) {
+        return new GenericSignCertificateProfileInfo(params);
+    }
+
+    /**
+     * Auth cert
+     *
+     * CN = serverCode
+     * C = country
+     * serialNumber = memberCode
+     * O = memberName
+     */
+    private static class GenericAuthCertificateProfileInfo extends AbstractCertificateProfileInfo
+            implements AuthCertificateProfileInfo {
+
+        GenericAuthCertificateProfileInfo(AuthCertificateProfileInfo.Parameters params) {
+            super(new DnFieldDescription[] {
+                new EnumLocalizedFieldDescriptionImpl("CN", DnFieldLabelLocalizationKey.SERVER_CODE,
+                        params.getServerId().getServerCode())
+                        .setReadOnly(true),
+                new EnumLocalizedFieldDescriptionImpl("serialNumber", DnFieldLabelLocalizationKey.MEMBER_CODE_SN,
+                        params.getServerId().getMemberCode())
+                        .setReadOnly(true),
+                new EnumLocalizedFieldDescriptionImpl("O", DnFieldLabelLocalizationKey.ORGANIZATION_NAME,
+                        params.getMemberName())
+                        .setReadOnly(true)}
+            );
+        }
+    }
+
+    /**
+     * Sign cert
+     *
+     * CN = memberName
+     * O = memberName
+     * businessCategory = memberClass
+     * C = country
+     * serialNumber = memberCode
+     */
+    private static class GenericSignCertificateProfileInfo extends AbstractCertificateProfileInfo
```

Comment:
The current implementation is missing instance identifier which makes it difficult to unambiguously map the signing certificate to a member. All fields in ClientID, country code and all fields in ServerId should be present in a generic profile. In that way, it's possible to identify the member who the certificate was issued to and the Security Server where it's used.

For example, using the Icelandic [sign cert profile](https://github.com/nordic-institute/X-Road/blob/develop/src/common-util/src/main/java/ee/ria/xroad/common/certificateprofile/impl/IsSignCertificateProfileInfo.java) and setting `setReadOnly(false)` for the country code would include all the necessary information.

Reply:
The instance id was left out on purpose. CAs do not have a specific field to set X-Road instance id on, like in the case of SK profile. We can not presume, that CAs will make X-Road compliant certificate profiles.

Another important reason is that not all X-Road instances are country based, there are multiple X-Road instances around the world that are used by private businesses, in such case, the instance ID is even harder to put somewhere in the certificate.

Reply:
As of such,  X-Road instance should not be mentioned in the certificate as the certification in a very standard approach is about the subject. The most important purpose of this change is to be able to use existing standard sign certification services on X-Road in small starting  (also non-governmental) X-Road instances.

Reply:
I agree that we cannot presume that CAs will make X-Road compliant certificate profiles. However, the `serialNumber` field is already used in the profile and instead of containing only the member code, it could contain other information as well, e.g., `ServerId` which includes the instance identifier too.

In the current implementation it's possible to have multiple certificates with the same `Subject Distinguished Name` value. When a member has a sign certificate on multiple Security Servers, all the sign certificates have the same `Subject Distinguished Name`. This could be an issue with some CAs and it could be avoided by including the server code in the profile.